### ISSUE TITLE
Can't pass -var to a compiled model

### ISSUE THREAD

**Issue by glucksfall** [Kappa-Dev/KaSim#523]
Hello everyone. 

I can no longer pass new variables to a model if it is compiled. For instance, consider the following model:

```
%agent: L(r)
%agent: R(l,r,Y~0~P)

%var: 'f' 1.000000e-02
%var: 'NA' 6.022140e+23
%var: 'EGF_conc_nM' 1.000000e+00
%var: 'KD1__FREE__' 1.000000e+00
%var: 'km1__FREE__' 1.000000e+00
%var: 'K2RT__FREE__' 1.000000e+00
%var: 'km2__FREE__' 1.000000e+00
%var: 'kphos__FREE__' 1.000000e+00
%var: 'kdephos__FREE__' 1.000000e+00
%var: 'zero' 0.000000e+00
%var: 'Vecf' 2.1e-9*'f'
%var: 'EGF_conc' 1.0e-9*'EGF_conc_nM'
%var: 'EGF_copy_number' 'EGF_conc'*'NA'*'Vecf'
%var: 'EGFR_copy_number' 100000.0*'f'
%var: 'KD1' 'KD1__FREE__'
%var: 'km1' 'km1__FREE__'
%var: 'kp1_nMs' 'km1'*'KD1'^(-1)
%var: 'kp1_Ms' 1000000000.0*'kp1_nMs'
%var: 'kp1' 'kp1_Ms'*('NA'*'Vecf')^(-1)
%var: 'K2RT' 'K2RT__FREE__'
%var: 'km2' 'km2__FREE__'
%var: 'kp2' 'K2RT'*'km2'*'EGFR_copy_number'^(-1)
%var: 'kp3' 0.2*'kp2'
%var: 'kphos' 'kphos__FREE__'
%var: 'kdephos' 'kdephos__FREE__'

'r1' L(r),R(l) -> L(r!1),R(l!1) @ 'kp1'
'r1_rev' L(r!1),R(l!1) -> L(r),R(l) @ 'km1'
'r2' R(l!_,r),R(l!_,r) -> R(l!_,r!1),R(l!_,r!1) @ 'kp2'
'r3' R(r!1),R(r!1) -> R(r),R(r) @ 'km2'
'r4' R(l,r),R(l,r) -> R(l,r!1),R(l,r!1) @ 'kp3'
'r5' R(r!_,Y~0) -> R(r!_,Y~P) @ 'kphos'
'r6' R(Y~P) -> R(Y~0) @ 'kdephos'

%obs: 'Ltot' |L()|
%obs: 'freeL' |L(r)|
%obs: 'Rtot' |R()|
%obs: 'RLbonds' |L(r!1),R(l!1)|
%obs: 'pR' |R(Y~P)|

%init: 'zero' L(r)
%init: 'EGFR_copy_number' R(l,r,Y~0)

%mod: alarm 1 do $PRINT <"">
%mod: alarm 1 do $PRINT <"kp1 is: " . 'kp1'>
%mod: alarm 1 do $PRINT <"km1 is: " . km1>
%mod: alarm 1 do $PRINT <"kp2 is: " . kp2>
%mod: alarm 1 do $PRINT <"km2 is: " . km2>
%mod: alarm 1 do $PRINT <"kp3 is: " . kp3>
%mod: alarm 1 do $PRINT <"kphos is: " . kphos>
%mod: alarm 1 do $PRINT <"kdephos is: " . kdephos>

%mod: alarm 600 do $ADD EGF_copy_number L(r)
```

And when using ```kasim4 -i model.kappa -l 660 -p 10 -o model_000_004.004.out.txt -syntax 3 -var KD1__FREE__ 0.029723 -var km1__FREE__ 3.628140 -var K2RT__FREE__ 0.037447 -var km2__FREE__ 60.072495 -var kphos__FREE__ 1.222907 -var kdephos__FREE__ 0.455667  -mode batch --no-log```, it prints the following to standard output:

```
kp1 is: 0.00965208756325
km1 is: 3.62814
kp2 is: 0.00224953472027
km2 is: 60.072495
kp3 is: 0.000449906944053
kphos is: 1.222907
kdephos is: 0.455667
```

while ```kasim4 -load-sim model.bin -l 660 -p 10 -o model_000_004.004.out.txt -syntax 3 -var KD1__FREE__ 0.029723 -var km1__FREE__ 3.628140 -var K2RT__FREE__ 0.037447 -var km2__FREE__ 60.072495 -var kphos__FREE__ 1.222907 -var kdephos__FREE__ 0.455667  -mode batch --no-log``` print the following

```
kp1 is: 7.90732988922e-05
km1 is: 1.
kp2 is: 0.001
km2 is: 1.
kp3 is: 0.0002
kphos is: 1.
kdephos is: 1.
```

The version of kasim4 is Kappa Simulator: v4.0rc1-244-gcaf2d9555-dirty (from GitHub).

Best regards.

**Comment by pirbo** [comment]
OK. This is """normal""" but extremely technical and I have no idea how to make things clearer.

TL;DR; don't use `-mode batch` when you `-make-sim model.bin`

 `-mode batch` means "I won't modify my model later". One of the consequences is to allow "constant propagation" (replacing in the model the variables by their values)

So when you say `./KaSim -make-sim model.bin models/523.ka -syntax 3 --compile -mode batch`, you end up with a compiled model containing
```
%var: 'km2__FREE__' 1.000000e+00
[...]
%var: 'km2' 1.0000000e+00
[...]
%mod: alarm 1 do $PRINT <"km2 is: " . 1.000000e+0>
```
instead of
```
%var: 'km2__FREE__' 1.000000e+00
[...]
%var: 'km2' 'km2__FREE__'
[...]
%mod: alarm 1 do $PRINT <"km2 is: " . km2>
```

Therefore when you later do `KaSim -load-sim model.bin -var km2__FREE__ 60.072495`, you obtain the model
```
%var: 'km2__FREE__' 60.072495
[...]
%var: 'km2' 1.0000000e+00
[...]
%mod: alarm 1 do $PRINT <"km2 is: " . 1.000000e+0>
```

If you would have done `./KaSim -make-sim model.bin models/523.ka -syntax 3 --compile`
then `./KaSim -load-sim model.bin -l 660 -p 10 -o model_000_004.004.out.txt -var KD1__FREE__ 0.029723 -var km1__FREE__ 3.628140 -var K2RT__FREE__ 0.037447 -var km2__FREE__ 60.072495 -var kphos__FREE__ 1.222907 -var kdephos__FREE__ 0.455667 -mode batch --no-log` would have did exactly what you expected...
